docs/contributing/999-rfcs/2024-06-02-ts-client.md

```diff
@@ -0,0 +1,121 @@
+---
+
+# TypeScript Clients Across Wing Applications
+
+- **Author(s)**: @MarkMcCulloh
+- **Submission Date**: 2024-06-02
+
+## Summary
+
+Allow for the generation of TypeScript clients that can interact with wing-defined infrastructure.
+Unlike `extern` or `@inflight`, the client is meant to be consumed by TypeScript applications rather than directly executed by Wing itself.
+
+The goal of this document is to walk through the DX of this experience.
+
+An important consideration for this discussion is allowing the generated client to reference resources defined from a separate wing application.
+To facilitate this separation, it seems necessary to introduce a mechanism to export/import deployed data between Wing apps.
+The rest of this proposal does it via `cloud.Outputs`, but that likely deserves its own RFC.
+
+## Example Situation
+
+Acme Co. has a platform team and several application teams. Their platform team manages Kubernetes clusters in AWS with additional AWS resources as needed.
+Application teams write applications in TypeScript and deploy them to these shared Kubernetes clusters.
+The centralized infrastructure definitions are written in Wing in their own repository.
+The applications are written in TypeScript but also have Wing files for infrastructure/deployment.
+
+To add a new application, the platform team needs to ensure that the application has the necessary permissions to interact with the shared resources. To do this, a kubernetes namespace and service account is created:
+
+```wing
+let namespace = cluster.createNamespace("app123");
+let serviceAccount = cluster.createServiceAccount(
+  namespace: namespace,
+  name: "app123-svc",
+);
+
+serviceAccount.grantRead(sharedBucket);
+```
+
+Applications using that namespace and service account will now have access to the shared bucket.
+
+To export this data to downstream applications, `cloud.Outputs` can be used:
+
+```wing
+let outputs = new cloud.Outputs();
+
+// Can reference preflight data (tokens will be resolved)
+outputs.add("namespace", namespace.name);
+outputs.add("serviceAccount", serviceAccount.name);
+
+// and inflight clients (includes the necessary information to interact with the resource for the given operations)
+// (e.g. a bucket name, a URL, etc.)
+outputs.lift("sharedBucket", sharedBucket, ["put"]);
+
+// Create a Wing library
+outputs.export(
+  name: "@acme/core-infra",
+);
+```
+
+This will create a centralized registry that can be imported elsewhere.
+During deployment, this registry will be updated with the relevant information.
+Additionally, it exports a wing library with all the necessary types and runtime code to look up that information.
+This library is published to an internal registry for other teams.
+
+After the central infrastructure is deployed, the application team can import the information into their own wing application:
+
+```wing
+bring "@acme/core-infra" as coreInfra;
+
+let workload = new Workload(
+  namespace: coreInfra.namespace,
```

Comment:
Let's assume wing is not used to define k8s resources.

Reply:
If wing is not used to deploy the application, how could the application know what information is needed?

If your answer is "the client", that's fine but the client itself still needs to look up the information. If that's the responsibility of the user, what help have we actually given them?

Reply:
It's a combination of "the client" and some "binding context" that is outputted by the deployment engine and somehow passed to the client upon initialization.